### PR TITLE
Handle unknown type in constraints

### PR DESCRIPTION
**Pull request by kaikalur**
Currently TypeConstraints abstraction is not used properly in SignatureBinder. So if we add a new constraint, that will not applied automatically and it needs to be duplicated in SignatureBinder. Upon close inspection, I see the only difference is in how it's handling UnknownType. So we move UnknownType to common.type (as it should be) and fix up the canBind logic to handle UnknownType properly.

Test plan - tests already exist.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-common/pom.xml [hunk 1]
```diff
@@ -91,5 +91,9 @@
             <artifactId>jackson-core</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
```
Comment (rongrong) [inline presto-common/pom.xml:94]: Remove this. I think it's no longer needed.

File: presto-common/src/main/java/com/facebook/presto/common/type/UnknownType.java [hunk 2]
```diff
@@ -130,7 +132,7 @@ public boolean getBoolean(Block block, int position)
     {
         // Ideally, this function should never be invoked for unknown type.
         // However, some logic rely on having a default value before the null check.
-        checkArgument(block.isNull(position));
+        checkArgument(block.isNull(position), "This function should never be called");
```
Comment (rongrong) [inline presto-common/src/main/java/com/facebook/presto/common/type/UnknownType.java:135]: The message, according to the check argument, should be "position %d is null in block". If this function (`getFixedSize`) should never be called, we should just throw `UnsupportedOperationException`. Otherwise the message is very confusing.

File: presto-common/src/main/java/com/facebook/presto/common/type/UnknownType.java [hunk 3]
```diff
@@ -140,7 +142,7 @@ public void writeBoolean(BlockBuilder blockBuilder, boolean value)
     {
         // Ideally, this function should never be invoked for unknown type.
         // However, some logic (e.g. AbstractMinMaxBy) rely on writing a default value before the null check.
-        checkArgument(!value);
+        checkArgument(!value, "This function should never be called");
```
Comment (rongrong) [inline presto-common/src/main/java/com/facebook/presto/common/type/UnknownType.java:145]: Same here.

File: presto-common/pom.xml (hunk 1 above)
Comment (kaikalur) [reply presto-common/pom.xml:94]: Hmm, I thought I removed it but the comment still exists?

Comment (kaikalur) [reply presto-common/pom.xml]: Done for real now :)

File: presto-common/src/main/java/com/facebook/presto/common/type/UnknownType.java (hunk 3 above)
Comment (kaikalur) [reply presto-common/src/main/java/com/facebook/presto/common/type/UnknownType.java]: Done

File: presto-common/src/main/java/com/facebook/presto/common/type/UnknownType.java (hunk 2 above)
Comment (kaikalur) [reply presto-common/src/main/java/com/facebook/presto/common/type/UnknownType.java]: Done

**Review by rongrong — COMMENTED** [review]
Please change the title to `Move UnknownType to presto-common`

File: presto-common/src/main/java/com/facebook/presto/common/type/UnknownType.java [hunk 4]
```diff
@@ -40,6 +35,13 @@ private UnknownType()
         super(new TypeSignature(NAME), boolean.class);
     }
 
+    private static void checkArgument(boolean condition, String message)
```
Comment (rongrong) [inline presto-common/src/main/java/com/facebook/presto/common/type/UnknownType.java:38]: Remove this?

File: presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java [hunk 5]
```diff
@@ -646,21 +634,7 @@ public SolverReturnStatus update(BoundVariables.Builder bindings)
...
-            if (!typeBound.isInstance(type)) {
-                return false;
-            }
-            if (requiredBaseName.isPresent() && !UNKNOWN.equals(type) && !requiredBaseName.get().equals(type.getTypeSignature().getBase())) {
-                // TODO: the case below should be properly handled:
-                // * `type` does not have the `requiredBaseName` but can be coerced to some type that has the `requiredBaseName`.
-                return false;
-            }
-            return true;
+            return typeVariableConstraint.canBind(type);
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:637]: Should just remove this private function and directly call `typeVariableConstraint.canBind`

File: presto-spi/src/main/java/com/facebook/presto/spi/function/TypeVariableConstraint.java [hunk 6]
```diff
@@ -91,12 +92,13 @@ public boolean canBind(Type type)
         if (orderableRequired && !type.isOrderable()) {
             return false;
         }
-        if (!typeBound.isInstance(type)) {
+        if (variadicBound != null && !UNKNOWN.equals(type) && !type.getTypeSignature().getBase().equals(variadicBound)) {
```
Comment (rongrong) [inline presto-spi/src/main/java/com/facebook/presto/spi/function/TypeVariableConstraint.java:95]: What's the reason of this change?

File: presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java (hunk 5 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:637]: The two function handles UNKNOWN differently. Is this change correct?

**Review by rongrong — COMMENTED** [review]
It's not clear to me why moving UnknownTyp to presto-common and the SignatureBinder refactoring should be in one commit. This makes naming the commit very hard...

File: presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java (hunk 5 above)
Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:637]: The code looks different but the logic is same (I think).

File: presto-common/src/main/java/com/facebook/presto/common/type/UnknownType.java (hunk 4 above)
Comment (kaikalur) [reply presto-common/src/main/java/com/facebook/presto/common/type/UnknownType.java:38]: Why? It's used!

File: presto-spi/src/main/java/com/facebook/presto/spi/function/TypeVariableConstraint.java (hunk 6 above)
Comment (kaikalur) [reply presto-spi/src/main/java/com/facebook/presto/spi/function/TypeVariableConstraint.java:95]: Because it was a hack due to UnknwonType not being available typeConstaint. So the main reason for moving UnknwonType is to be able to use it in this classs.

**Comment by kaikalur** [comment]
> It's not clear to me why moving UnknownTyp to presto-common and the SignatureBinder refactoring should be in one commit. This makes naming the commit very hard...

See my other comment. Though it looks like it's a good thing to move UnknownType to common.type, for now the only use is in the TypeConstraint. So I just went from that angle.

File: presto-common/src/main/java/com/facebook/presto/common/type/UnknownType.java (hunk 4 above)
Comment (rongrong) [reply presto-common/src/main/java/com/facebook/presto/common/type/UnknownType.java:38]: not in the diff so i didn't see it 😂

File: presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java (hunk 5 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:637]: Can you run a verifier explain test to double check on this?

File: presto-spi/src/main/java/com/facebook/presto/spi/function/TypeVariableConstraint.java (hunk 6 above)
Comment (rongrong) [reply presto-spi/src/main/java/com/facebook/presto/spi/function/TypeVariableConstraint.java:96]: Change the commit message to reflect this then.

**Review by rongrong — APPROVED** [review]
Looks good. Please change the commit message to capture the logic change as well. Also I'd prefer to do a round of explain test to make sure the signature binder change is safe.

**Comment by kaikalur** [comment]
> Looks good. Please change the commit message to capture the logic change as well. Also I'd prefer to do a round of explain test to make sure the signature binder change is safe.

Explain tests completed with no errors.

File: presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java (hunk 5 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:637]: Can we delete this function now?

Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java:637]: I thought I did !

Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java]: Done for real now :)

File: presto-spi/src/main/java/com/facebook/presto/spi/function/TypeVariableConstraint.java (hunk 6 above)
Comment (kaikalur) [reply presto-spi/src/main/java/com/facebook/presto/spi/function/TypeVariableConstraint.java]: Done
